packages/mui-material/src/Button/Button.js

```diff
@@ -343,8 +404,131 @@ const ButtonEndIcon = styled('span', {
         marginRight: -2,
       },
     },
+    {
+      props: { loadingPosition: 'end', loading: true },
+      style: {
+        transition: theme.transitions.create(['opacity'], {
+          duration: theme.transitions.duration.short,
+        }),
+        opacity: 0,
+      },
+    },
+    {
+      props: ({ ownerState }) => ownerState.loadingPosition === 'end' && ownerState.fullWidth,
+      style: {
+        transition: theme.transitions.create(['opacity'], {
+          duration: theme.transitions.duration.short,
+        }),
+        opacity: 0,
+        marginLeft: -8,
+      },
+    },
     ...commonIconStyles,
   ],
+}));
+
+const ButtonLoadingIndicator = styled('span', {
+  name: 'MuiButton',
+  slot: 'LoadingIndicator',
+  overridesResolver: (props, styles) => {
+    const { ownerState } = props;
+    return [
+      styles.loadingIndicator,
+      styles[`loadingIndicator${capitalize(ownerState.loadingPosition)}`],
+    ];
+  },
+})(({ theme }) => ({
+  position: 'absolute',
+  visibility: 'visible',
+  display: 'flex',
+  variants: [
+    {
+      props: {
+        loadingPosition: 'start',
+        size: 'small',
+      },
+      style: {
+        left: 10,
+      },
+    },
+    {
+      props: ({ loadingPosition, ownerState }) =>
+        loadingPosition === 'start' && ownerState.size !== 'small',
+      style: {
+        left: 14,
+      },
+    },
+    {
+      props: {
+        variant: 'text',
+        loadingPosition: 'start',
+      },
+      style: {
+        left: 6,
+      },
+    },
+    {
+      props: {
+        loadingPosition: 'center',
+      },
+      style: {
+        left: '50%',
+        transform: 'translate(-50%)',
+        color: (theme.vars || theme).palette.action.disabled,
+      },
+    },
+    {
+      props: {
+        loadingPosition: 'end',
+        size: 'small',
+      },
+      style: {
+        right: 10,
+      },
+    },
+    {
+      props: ({ loadingPosition, ownerState }) =>
+        loadingPosition === 'end' && ownerState.size !== 'small',
+      style: {
+        right: 14,
+      },
+    },
+    {
+      props: {
+        variant: 'text',
+        loadingPosition: 'end',
+      },
+      style: {
+        right: 6,
+      },
+    },
+    {
+      props: ({ ownerState }) => ownerState.loadingPosition === 'start' && ownerState.fullWidth,
+      style: {
+        position: 'relative',
+        left: -10,
+      },
+    },
+    {
+      props: ({ ownerState }) => ownerState.loadingPosition === 'end' && ownerState.fullWidth,
+      style: {
+        position: 'relative',
+        right: -10,
+      },
+    },
+  ],
+}));
+
+const ButtonLoadingLabel = styled('span', {
+  name: 'MuiButton',
+  slot: 'Label',
+  overridesResolver: (props, styles) => {
+    return [styles.label];
+  },
+})({
+  display: 'inherit',
+  alignItems: 'inherit',
+  justifyContent: 'inherit',
```

Comment:
This seems too magical. Is there a way we could implement the `loading` prop without adding a wrapper around `children`? I guess the answer is no, but just in case.

Reply:
The `span` was added in #35198 and the styles around it was added in #43436

Reply:
@siriwatknp Sure